src/hotspot/share/runtime/interfaceSupport.inline.hpp

```diff
@@ -158,7 +143,8 @@ class ThreadInVMForHandshake : public ThreadStateTransition {
   }
 
   ~ThreadInVMForHandshake() {
-    transition_back();
+    assert(_thread->thread_state() == _thread_in_vm, "should only call when leaving VM after handshake");
```

Comment:
Can we also assert `!_thread->has_special_runtime_exit_condition()`? Or at least that the value of this has not changed between the TIVFH constructor and destructor.

Reply:
has_special_runtime_exit_condition() checks for async exceptions, external_suspend and JFR suspend. The last two can actually be set while we are in the handshake. It doesn't mean that the caller will consider the target suspended, but the _suspend_flags might be different at construction and destruction of TIVFH. As for async exceptions, I realized we have a similar issue as the one described in 8223572 when calling handle_polling_page_exception() on a return poll:

- T calls ThreadSafepointState::handle_polling_page_exception() because another thread called T.stop() (which we implement with a handshake)
- T calls SafepointMechanism::process_if_requested(thread())
- T calls HandshakeState::process_self_inner() and processes the handshake which calls set_pending_async_exception()
- T returns from SafepointMechanism::process_if_requested(thread()) and goes back to java without calling check_and_handle_async_exceptions() to throw the ThreadDeath() exception

I'm not sure if delaying the throw of ThreadDeath is allowed, so if we need to keep the current behaviour I can add a call to check_and_handle_async_exceptions():

@@ -987,6 +987,7 @@ void ThreadSafepointState::handle_polling_page_exception() {
     // Process pending operation
     SafepointMechanism::process_if_requested(thread());
+   thread()->check_and_handle_async_exceptions();

     // restore oop result, if any

Reply:
Reading that comment I realized that this change would break in flight pr #119 

To refresh memories: there I'm adding a new suspend flag currently required for object reallocation. I'm using a handshake to synchronize with the suspendee thread S. After the handshake it is guaranteed that the stack of S is walkable and S cannot push/pop frames. This would not work anymore.

I could replace the handshakes (in [escapeBarrier.cpp line 172, 218](https://github.com/openjdk/jdk/pull/119/files#diff-b72c15f49cba82da2158d4e9a4ab0e92R172-R173)) with vm operations though. Should be ok.

Reply:
I see. You should still be able to use handshakes. I think the issue is that in handle_polling_page_exception(), instead of calling SafepointMechanism::process_if_requested(thread()), we should be doing { ThreadInVMfromJava __tiv(thread());} and { ThreadInVMfromJavaNoAsyncException __tiv(thread());} for the polling on return and not on return respectively. That will process a safepoint/handshake and then check the special conditions before going back to java.

Reply:
I thought about removing that check, but right now it's also preventing an escape from a suspend. The issue is that when going back to java, we always check if we were suspended and call java_suspend_self_with_safepoint_check() if true. After the do-while(is_external_suspend()) loop we call again SafepointMechanism::process_if_requested(this) to see if we need to stop for a safepoint/handshake. If there is an actual safepoint pending because of a VM_ThreadSuspend() then unless we keep that check the thread will not notice it was suspended after the safepoint and will go back to java.
Now, I'm thinking the call to SafepointMechanism::process_if_requested(this) in java_suspend_self_with_safepoint_check() should be moved inside the the do-while(is_external_suspend()). That way we guarantee after returning from java_suspend_self_with_safepoint_check() that no one will be assuming the thread is suspended. And in that case I think that check in SS::block can be removed.

Reply:
> I thought about removing that check, but right now it's also preventing an
> escape from a suspend. The issue is that when going back to java, we always
> check if we were suspended and call java_suspend_self_with_safepoint_check()
> if true. After the do-while(is_external_suspend()) loop we call again
> SafepointMechanism::process_if_requested(this) to see if we need to stop for a
> safepoint/handshake. If there is an actual safepoint pending because of a
> VM_ThreadSuspend() then unless we keep that check the thread will not notice
> it was suspended after the safepoint and will go back to java.

I meant the block could be removed if we put
ThreadInVMfromJava/ThreadInVMfromJava (TIVFJ) into
handle_polling_page_exception(). In that case the thread would notice and honor
the suspend when leaving the TIVFJ block. This would not be too late, would it?

> Now, I'm thinking the call to SafepointMechanism::process_if_requested(this)
> in java_suspend_self_with_safepoint_check() should be moved inside the the
> do-while(is_external_suspend()). That way we guarantee after returning from
> java_suspend_self_with_safepoint_check() that no one will be assuming the
> thread is suspended. And in that case I think that check in SS::block can be
> removed.

Yes I agree.

Well, I don't want to spam your PR with too many questions / comments. The
change still looks good to me.

Reply:
> I meant the block could be removed if we put
> ThreadInVMfromJava/ThreadInVMfromJava (TIVFJ) into
> handle_polling_page_exception(). In that case the thread would notice and honor
> the suspend when leaving the TIVFJ block. This would not be too late, would it?

Yes, it will honor the first suspend in the TIVFJ destructor by calling handle_special_runtime_exit_condition(), but even in that case, we would still need the check in SS:block() because of the last call to SafepointMechanism::process_if_requested(thread()) in java_suspend_self_with_safepoint_check(). That could be a new VM_ThreadSuspend() operation to suspend the thread again. So that check in SS:block() is actually needed for all the transitions to java not only the one in handle_polling_page_exception(). But I think we can remove that recursion if we move that call to SafepointMechanism::process_if_requested(thread()) inside the do-while().

Reply:
> > Now, I'm thinking the call to SafepointMechanism::process_if_requested(this)
> > in java_suspend_self_with_safepoint_check() should be moved inside the the
> > do-while(is_external_suspend()). That way we guarantee after returning from
> > java_suspend_self_with_safepoint_check() that no one will be assuming the
> > thread is suspended. And in that case I think that check in SS::block can be
> > removed.
> 
> Yes I agree.
Great. I'll test it out.

> Well, I don't want to spam your PR with too many questions / comments. The
> change still looks good to me.
No problem, it's actually good to have more eyes looking to make sure I don't miss anything.  : )
Thanks Richard!

Reply:
> > I meant the block could be removed if we put
> > ThreadInVMfromJava/ThreadInVMfromJava (TIVFJ) into
> > handle_polling_page_exception(). In that case the thread would notice and honor
> > the suspend when leaving the TIVFJ block. This would not be too late, would it?
> 

> Yes, it will honor the first suspend in the TIVFJ destructor by calling
> handle_special_runtime_exit_condition(), but even in that case, we would still
> need the check in SS:block() because of the last call to
> SafepointMechanism::process_if_requested(thread()) in
> java_suspend_self_with_safepoint_check(). That could be a new
> VM_ThreadSuspend() operation to suspend the thread again.

Sorry, I have to ask again to really understand this myself. If it is a new
VM_ThreadSuspend() that started while the target thread T was still
_thread_blocked in java_suspend_self_with_safepoint_check(), then T can not
leave the do-while-loop because is_external_suspend() will return true as
setting the suspend flag for T cannot be reordered with the begin of the
safepoint.

Also if T could leave the loop then the safepoint could be ended before T
reaches SS::block(). It would escape then to java even though it was suspended.

See also review thread for JDK-8252521:
http://mail.openjdk.java.net/pipermail/hotspot-runtime-dev/2020-September/041632.html

Reply:
> > > I meant the block could be removed if we put
> > > ThreadInVMfromJava/ThreadInVMfromJava (TIVFJ) into
> > > handle_polling_page_exception(). In that case the thread would notice and honor
> > > the suspend when leaving the TIVFJ block. This would not be too late, would it?
> 
> > Yes, it will honor the first suspend in the TIVFJ destructor by calling
> > handle_special_runtime_exit_condition(), but even in that case, we would still
> > need the check in SS:block() because of the last call to
> > SafepointMechanism::process_if_requested(thread()) in
> > java_suspend_self_with_safepoint_check(). That could be a new
> > VM_ThreadSuspend() operation to suspend the thread again.
> 
> Sorry, I have to ask again to really understand this myself. If it is a new
> VM_ThreadSuspend() that started while the target thread T was still
> _thread_blocked in java_suspend_self_with_safepoint_check(), then T can not
> leave the do-while-loop because is_external_suspend() will return true as
> setting the suspend flag for T cannot be reordered with the begin of the
> safepoint.
The new JVM_SuspendThread() call (up to VM_ThreadSuspend()) is started after the thread sets its state back to _thread_in_Java and exits the do-while() loop but before it calls SafepointMechanism::process_if_requested(this). The thread will see there is a safepoint pending and will block in SS::block(). If there is no check in SS::block(), after the safepoint finishes the thread will return to java_suspend_self_with_safepoint_check(), back to ~ TIVFJ and then back to java. 

> Also if T could leave the loop then the safepoint could be ended before T
> reaches SS::block(). It would escape then to java even though it was suspended.
If the VMThread doesn't read the thread's state before it is restored back to _thread_in_Java then the safepoint won't progress since the thread is unsafe (_thread_in_Java). So it will have to wait for the thread to call SS::block(). If the VMThread did read a safe state (_thread_blocked) then it means we will see we were suspended in the do-while() loop and suspend.

Reply:
Understood now. Thanks!